learntools/core/problem_view.py

```diff
@@ -43,6 +45,28 @@ def __getattr__(self, attr):
             return val
         raise AttributeError
     
+    @property
+    def questionId(self):
+        # e.g. '3_MyHardProblem'
+        id = self.problem.__class__.__name__
+        if hasattr(self, '_order'):
```

Comment:
I think of the problem ID and where it falls in the exercises as independent. 

There are exceptions (e.g. we have reason to think a problem becomes easier or harder based on what precedes it.

But I think that will be the exception rather than the rule. So I'd rather place the burden on the exercise writer to change the exercise name when that happens, rather than requiring the analyst to munge out problem names in the query.

Reply:
My thinking here was 

1. If I'm looking at an exercise notebook, it's a little bit of effort to align the questions with questionIds in the db - because the class name only occurs in the source code. But if the questionId includes the question number, it becomes really easy.
2. It makes it easier to do a survival analysis. e.g. how many users tried questions 1 and 2 and quit at question 3?

I think the main downside is questionIds shifting if we add/remove/reorder questions within an existing exercise. 

I'm a little on the fence about it, but leaning towards thinking the benefits outweigh that downside, since changes like that seem very rare.

(I agree that ideally the order *should* be independent - I just wish I had thought of it back when we were coming up with the schema.)

Reply:
I **really** don't want this to become heavier weight than it ought to be... but this also tempts me into thinking the correspondence (what exercises are in what order in what notebook) should be stored in some metadata file... and not anywhere else in the codebase.

Thoughts?

Reply:
^ that also maybe being closer to a path to someday integrating into the content API, so that the tutorialIds are inserted programmatically rather than us hard-coding them in these python files.

Reply:
The ordering of problems is currently only stored in one place in the code base, so I don't think this would be any more DRY-y. Also I don't think it results in a clean separation of abstractions. That metadata file is presumably going to be referring to Python class names, so I think we might as well define the ordering in the same module where we're defining those classes.

Reply:
But the current solution requires we embed some logic in the analytics to remove the ordering.  

The argument for keeping it this way was:
1) Align questions to exercises
2) Facilitate survival analysis

I'm claiming we could do these programmatically if the ordering was stored as metadata... then the ordering doesn't need to be appended to the problem name in the DB, and the name can refer directly to the problem.

Reply:
Sorry... Thought I'd posted this before you took a look.  meant to include this when first commenting... but copying it here now:

---

Since we already have a concept of point_value, one way I could see this working is giving each problem a `points_multiplier` property set to 1 by default.

The problem author could reset `points_multiplier` as part of `check`, creating a score that's continuous.  Then the line above becomes
kwargs['valueTowardsCompletion'] = self.problem.point_value * self.points_multiplier

Maybe there's a better process for how `points_multiplier` is updated. But I don't think it requires a breaking change to the current abstractions.

Reply:
> I'm claiming we could do these programmatically if the ordering was stored as metadata... 

Ah, okay. In practice, what would that look like? Let's say we have JSON files (per track? per exercise?) with the order of problems within exercises. How would we go about using that for analysis? Maybe we write a Python script that takes one of these files and outputs some chunk of SQL creating a temporary table that puts questionIds and ordinals side by side? Then when we want to analyze the interaction data, we run this script, copy the "WITH (..." string into SQL explorer, and then write our query?

That feels pretty clunky, but maybe you were imagining something different?

> Maybe there's a better process for how points_multiplier is updated. But I don't think it requires a breaking change to the current abstractions.

For me the disconnect is that I'm imagining the intended purposes of the ValueTowardsCompletion field is determining whether a user has completed an exercise (i.e. whether they get a checkbox, or how much to fill in their progress bar). 

It sounds like what you're describing with the ticket pricing example is about collecting some data that's interesting from an analytics point of view, but not intended to be used to determine whether an exercise is done.

(I suppose you could say that, for this question, the target MSE is 80. So if your model has a MSE of 160, you're 50% done. But would you actually want to surface it that way in the UI? I think this would be especially confusing if this kind of scoring is combined with other questions in the same exercise. If I do really well on question 1 and do 4x better than the threshold, does that result in me getting a checkmark for the whole exercise, without touching questions 2, 3, and 4?)

Reply:
**That feels pretty clunky, but maybe you were imagining something different?**

We can access SQL explorer from local jupyter notebooks.  So I imagine pulling all the data we need from SQL explorer using a query, but then locally loading a metadata file... and then doing the appropriate merge in pandas and doing everything else in pandas.

I'm not totally convinced about the metadata file idea.  But no matter what we decide on that, I think we'll want to create the analytics in Python rather than SQL (and I should help you get set up to remotely query the DB, which I found somewhat painful when I first set it up).

I'm not totally sold on any of the approaches we've discussed. But if I had to choose right now, my inclination would be to pull out the stuff about `_order` right now. And feel better about it knowing that if we find we need to do things like survival analysis where that would be useful, we can achieve the same things by storing that data in a separate file rather than encoding it in the problem name.

Reply:
I'm going to keep this as is for a moment, but I'm definitely open to changing it in the future (and I think we can do so without much fallout).